### PR TITLE
Optimize if function when input type is map

### PR DESCRIPTION
**Pull request by taiyang-li**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Optimize if function when input type is map, speed up by ~10x

``` sql 
with rand32() % 2 as x select if(x, map(1,2,3,4), map(3,4,5,6)) from numbers(10000000) format Null

Old: 0 rows in set. Elapsed: 4.293 sec. Processed 10.00 million rows, 80.00 MB (2.33 million rows/s., 18.63 MB/s.)
New: 0 rows in set. Elapsed: 0.411 sec. Processed 10.00 million rows, 80.00 MB (24.33 million rows/s., 194.64 MB/s.)
```

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit dc7b9d8208c9d696c08c6c8aa5332054b187ca7e with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59413/dc7b9d8208c9d696c08c6c8aa5332054b187ca7e/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59413/dc7b9d8208c9d696c08c6c8aa5332054b187ca7e/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Functions/if.cpp [hunk 1]
```diff
@@ -679,6 +681,87 @@ class FunctionIf : public FunctionIfBase
...
+        const DataTypeMap & map_result_type = static_cast<const DataTypeMap &>(*result_type);
+        auto key_result_type = std::make_shared<DataTypeArray>(map_result_type.getKeyType());
+        auto value_result_type = std::make_shared<DataTypeArray>(map_result_type.getValueType());
+        ColumnPtr key_result = executeImpl(key_columns, key_result_type, input_rows_count);
+        ColumnPtr value_result = executeImpl(value_columns, value_result_type, input_rows_count);
+
+        /// key_result and value_result are not constant columns otherwise we won't reach here in executeMap
+        const auto * key_array = assert_cast<const ColumnArray *>(key_result.get());
+        const auto * value_array = assert_cast<const ColumnArray *>(value_result.get());
+        if (!key_array)
```
Comment (Algunenano) [inline src/Functions/if.cpp:745]: Should all 3 of these exceptions be LOGICAL_ERROR?

File: src/Functions/if.cpp [hunk 2]
```diff
@@ -1,33 +1,34 @@
-#include <DataTypes/DataTypesNumber.h>
-#include <DataTypes/DataTypesDecimal.h>
+#include <type_traits>
```
Comment (Algunenano) [inline src/Functions/if.cpp:1]: Personal preference, but I prefer keeping std headers in a different block and after project headers.

File: tests/queries/0_stateless/02974_if_with_map.sql [hunk 3]
```diff
@@ -0,0 +1,15 @@
+select if(number % 2 = 0, map(1,2,3,4), map(3,4,5,6)) from numbers(2);
+select if(number % 2 = 0, materialize(map(1,2,3,4)), map(3,4,5,6)) from numbers(2);
+select if(number % 2 = 0, map(3,4,5,6), materialize(map(1,2,3,4))) from numbers(2);
```
Comment (Algunenano) [inline tests/queries/0_stateless/02974_if_with_map.sql:3]: Let's add a similar test test with maps of different sizes

File: src/Functions/if.cpp (hunk 2 above)
Comment (taiyang-li) [reply src/Functions/if.cpp:1]: It's the formatted result of clang-format. I'll seperate std headers in another group

File: src/Functions/if.cpp (hunk 1 above)
Comment (taiyang-li) [reply src/Functions/if.cpp:745]: fixed

**Review by Algunenano — APPROVED** [review]
Needs style fix, but looks great to me.

**Comment by taiyang-li** [comment]
> Needs style fix, but looks great to me.

Done.

**Comment by Algunenano** [comment]
02906_orc_tuple_field_prune failure is related. Please have a look and add a test once you find a smaller reproducer

**Review by Algunenano — APPROVED** [review]

**Pull request merged**